libpromises/policyserver.c

```diff
@@ -0,0 +1,300 @@
+}
+
+const char* GetPolicyServerIP()
+{
+    int ret = Hostname2IPString(POLICY_SERVER_IP, POLICY_SERVER_HOST, CF_MAX_IP_LEN);
+    if (ret != 0)
+    {
+        return NULL;
+    }
+    return PolicyServerCheckedReturn(POLICY_SERVER_IP);
+}
+
+const char* GetPolicyServerHost()
+{
+    return PolicyServerCheckedReturn(POLICY_SERVER_HOST);
+}
+
+const char* GetPolicyServerPort()
+{
+    return PolicyServerCheckedReturn(POLICY_SERVER_PORT);
+}
+
+
+//*******************************************************************
+// POLICY SERVER FILE FUNCTIONS:
+//*******************************************************************
+
+static char *PolicyServerFilename(const char *workdir)
+{
+    return StringFormat("%s%cpolicy_server.dat", workdir, FILE_SEPARATOR);
+}
+
+char *ReadPolicyServerFile(const char *workdir)
+{
+    char contents[CF_MAX_IP_LEN] = "";
+
+    char *filename = PolicyServerFilename(workdir);
+    FILE *fp = fopen(filename, "r");
+
+    if (fp)
+    {
+        if (fscanf(fp, "%63s", contents) != 1)
+        {
+            fclose(fp);
+            return NULL;
+        }
+        fclose(fp);
+        free(filename);
+        return xstrdup(contents);
+    }
+    else
+    {
+        Log( LOG_LEVEL_VERBOSE, "Could not open file '%s' (fopen: %s)",
+             filename, GetErrorStr() );
+
+        free(filename);
+        return NULL;
+    }
+}
+
+void SetPolicyServerFromFile(EvalContext *ctx, const char* workdir)
+{
+    char* contents = ReadPolicyServerFile(workdir);
+    SetPolicyServer(ctx, contents);
+    free(contents);
+}
+
+bool ParsePolicyServerFile(const char *workdir, char **host, char **port)
+{
+    char* contents = ReadPolicyServerFile(workdir);
+    if (contents == NULL)
+    {
+        return false;
+    }
+    (*host) = NULL;
+    (*port) = NULL;
+
+    ParseHostPort(contents, host, port);
+    (*host) = xstrdup(*host);
+    if (*port != NULL)
+    {
+        (*port) = xstrdup(*port);
+    }
+    free(contents);
+    return true;
+}
+
+bool LookUpPolicyServerFile(const char *workdir, char **ipaddr, char **port)
+{
+    char *host;
+    bool file_read = ParsePolicyServerFile(workdir, &host, port);
+    if (file_read == false)
+    {
+        return false;
+    }
+    char tmp_ipaddr[CF_MAX_IP_LEN];
+    if (Hostname2IPString(tmp_ipaddr, host, sizeof(tmp_ipaddr)) == -1)
+    {
+        Log(LOG_LEVEL_ERR,
+            "Unable to resolve policy server host: %s", host);
+        return false;
+    }
+    (*ipaddr) = xstrdup(tmp_ipaddr);
+    free(host);
+    return true;
+}
+
+bool WritePolicyServerFile(const char *workdir, const char *new_policy_server)
+{
+    char *filename = PolicyServerFilename(workdir);
+
+    FILE *file = fopen(filename, "w");
+    if (!file)
+    {
+        Log(LOG_LEVEL_ERR, "Unable to write policy server file '%s' (fopen: %s)", filename, GetErrorStr());
+        free(filename);
+        return false;
+    }
+
+    fprintf(file, "%s", new_policy_server);
```

Comment:
Since you are at it, would you mind adding a trailing `\n` in a separate commit at the end? I don't think it will break existing code, and I've been wanting to do it a since a long time...